### PR TITLE
Add UEFI Get Variable support

### PR DESCRIPTION
**Pull request by stalamudupula**
This patch provides a library to get read access
to the UEFI variables stored in NV region.

Signed-off-by: Sai Talamudupula <sai.kiran.talamudupula@intel.com>

### CODE REVIEW

**Comment by mauricema** [comment]
For the commit message, can we provide a little bit more background on where is the driver ported form, and why it was added ?

File: Platform/CoffeelakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c [hunk 1]
```diff
@@ -868,6 +868,7 @@ BoardInit (
   EFI_PEI_GRAPHICS_INFO_HOB *FspGfxHob;
   LOADER_GLOBAL_DATA        *LdrGlobal;
 
+
```
Comment (mauricema) [inline Platform/CoffeelakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c]: Remove empty line

File: Platform/CoffeelakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c [hunk 2]
```diff
@@ -732,6 +733,7 @@ DEBUG_CODE_END();
     PlatformFeaturesInit ();
     break;
   case PreMemoryInit:
+    break;
```
Comment (mauricema) [inline Platform/CoffeelakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: why do you add "break" here?   I think the intention was to call GpioInit() for both PreMemoryInit and PostMemoryInit.

File: Platform/CoffeelakeBoardPkg/CfgData/CfgDataInt_Cflh.dlt [hunk 3]
```diff
@@ -38,6 +38,7 @@ MEMORY_CFG_DATA.UserBd                    | 0x01
 MEMORY_CFG_DATA.DualDpc                   | 0x01
 
 GPU_CFG_DATA.DdiPortFHpd                  | 0x0
+GEN_CFG_DATA.PayloadId                    | 'UEFI'
```
Comment (mauricema) [inline Platform/CoffeelakeBoardPkg/CfgData/CfgDataInt_Cflh.dlt]: Why do we change this default PayloadId ?   It is not related to the UEFI variable.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  gEfiVariableIndexTableGuid
+  gEfiSystemNvDataFvGuid            ## SOMETIMES_CONSUMES   ## GUID
+  ## SOMETIMES_CONSUMES   ## HOB
+  ## CONSUMES             ## GUID # Dependence
+  gEdkiiFaultTolerantWriteGuid
+  gBootLoaderCombinedVariableHdrGuid
+  gEdkiiWorkingBlockSignatureGuid
+
+[Ppis]
+  gEfiPeiReadOnlyVariable2PpiGuid   ## PRODUCES
```
Comment (mauricema) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf]: Do we use this PPI ?

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+[Pcd]
+  gPlatformCommonLibTokenSpaceGuid.PcdUefiVariableLibId
+
+[Depex]
+  gEdkiiFaultTolerantWriteGuid
```
Comment (mauricema) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf]: No point to put dependency here for library.  SBL does not use it.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c [hunk 6]
```diff
@@ -0,0 +1,1055 @@
...
+      //
+      return FALSE;
+    }
+    if (((UINTN) Variable < (UINTN) TargetAddress) && (((UINTN) Variable + GetVariableHeaderSize (StoreInfo->AuthFlag)) > (UINTN) TargetAddress)) {
+      //
+      // Variable header pointed by Variable is inconsecutive,
+      //
+      LibStatus = GetLibraryData (PcdGet8 (PcdUefiVariableLibId), (VOID **)&VarStoreLibData);
+      if (LibStatus == EFI_NOT_FOUND) {
+        *VariableHeader = &VarStoreLibData->VariableHdr;
```
Comment (mauricema) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c]: Will we get "EFI_NOT_FOUND"?  The library data should have been set previously.  So it should always return success.  I think the logic here needs adjustment.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  ## SOMETIMES_CONSUMES   ## HOB
+  gEfiVariableGuid
+  ## SOMETIMES_PRODUCES   ## HOB
+  ## SOMETIMES_CONSUMES   ## HOB
+  gEfiVariableIndexTableGuid
+  gEfiSystemNvDataFvGuid            ## SOMETIMES_CONSUMES   ## GUID
+  ## SOMETIMES_CONSUMES   ## HOB
+  ## CONSUMES             ## GUID # Dependence
+  gEdkiiFaultTolerantWriteGuid
+  gBootLoaderCombinedVariableHdrGuid
```
Comment (mauricema) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf]: is this used?

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  ## SOMETIMES_CONSUMES   ## GUID # Variable store header
+  ## SOMETIMES_CONSUMES   ## HOB
+  gEfiVariableGuid
+  ## SOMETIMES_PRODUCES   ## HOB
+  ## SOMETIMES_CONSUMES   ## HOB
+  gEfiVariableIndexTableGuid
+  gEfiSystemNvDataFvGuid            ## SOMETIMES_CONSUMES   ## GUID
+  ## SOMETIMES_CONSUMES   ## HOB
+  ## CONSUMES             ## GUID # Dependence
+  gEdkiiFaultTolerantWriteGuid
```
Comment (mauricema) [inline BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf]: is this used?

File: Platform/CoffeelakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c (hunk 2 above)
Comment (stalamudupula) [reply Platform/CoffeelakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: True. Corrected.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf (hunk 4 above)
Comment (stalamudupula) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf]: Right. Deleted.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf (hunk 5 above)
Comment (stalamudupula) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.inf]: Deleted.

File: BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c (hunk 6 above)
Comment (stalamudupula) [reply BootloaderCommonPkg/Library/UefiVariableLib/UefiVariableLib.c]: Used a flag to indicate if VariableHdr is updated in library data. If not, only then take this path.

**Pull request merged**